syntax = "proto2";

package bismuth;


// For POC, address are strings, but they should be bytes of fixed size, half the size.
message TX {
    // Timestamp
    required uint32 ts=1;
    // Who - tester
    required string from=2;
    // testee
    required string to=3;
    // Tx or test type
    required uint32 type=4;
    // params of the test
    optional string params=5;
    // results of the test
    required string result=6;
    // signature = txid
    required string signature=7;
}

message Block {
    required uint32 height=1;
    required uint32 round=2;
    required uint32 sir=3;
    // Timestamp
    required uint32 ts=4;
    required string previous_hash=5;
    repeated TX txs = 6;
    required uint32 msg_count=7;
    required uint32 sources=8;
    // Who - forger
    required string from=9;
    // Should be byte too
    required string signature=10;
    // Blockhash
    required string hash=11;
}

message IP {
    // Would "bytes" be better suited?
    required uint32 ipv4=1;
    optional uint64 ipv6a=2;
    optional uint64 ipv6b=3;
    optional uint32 port=4 [default = 5658 ];
}

enum ko_reason {
    no_reason = 0;
    version_mismatch = 1;
    bad_ip = 2;
    bad_pubkey = 3;
    bad_block = 4;
    bad_slot = 5;
    no_resources = 6;
}


message Command {
    enum Type {
        hello = 0;
        ok = 1;
        ko = 2;
        ping = 3;
        peers = 4;
        status = 5;
        block = 6;
        txsigs = 7;
        txs = 8;
        //message = 10;
    }
    required Type command = 1;

    oneof param {
        string string_value=2;
        int32 int32_value=3;
        //ko_reason reason_value=4;
    }

    repeated TX tx_values = 5;
    repeated IP ip_values = 6;
    repeated string sig_values = 7;
    optional Block block_value = 8;
}
